File: Week 2/PSet2/readability/readability.c
#include <ctype.h>
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
float get_100_word_avg(int count, int nr_of_words);
int get_grade(int letter_count, int sentence_count, int word_count);

int main(void)
{
    // Get user input
    string text = get_string("Text: ");
    // Get letters count
    int letter_count = count_letters(text);
    // Get words count
    int word_count = count_words(text);
    // Get sentence count
    int sentence_count = count_sentences(text);
    // Calculate grade
    int grade = get_grade(letter_count, sentence_count, word_count);
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
    // printf("%i %i %i Grade: %i\n", letter_count, word_count, sentence_count, grade);
}

int count_letters(string text)
{
    int letter_count = 0;
    // Loop through text
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        char text_element = text[i];
        // Increase count when char is letter
        if (isalpha(text_element))
        {
            letter_count++;
        }
    }
    return letter_count;
}

int count_words(string text)
{
    int word_count = 0;
    // Loop through text
    for (int i = 0, n = strlen(text); i <= n; i++)
    {
        char text_element = text[i];
        // Check if char is space or null
        if (isspace(text_element) || i == n)
        {
            word_count ++;
        }
    }
    return word_count;
}

int count_sentences(string text)
{
    int sentence_count = 0;
    // Loop through text
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        char text_element = text[i];
        // Check if character is punctuation
        if (text_element == '!' || text_element == '.' || text_element == '?')
        {
            sentence_count ++;
        }
    }
    return sentence_count;
}

float get_100_word_avg(int count, int nr_of_words)
{
    return count / (float)nr_of_words * 100.0;
}

int get_grade(int letter_count, int sentence_count, int word_count)
{
    return round(0.0588 * get_100_word_avg(letter_count, word_count) - 0.296 * get_100_word_avg(sentence_count, word_count) - 15.8);
}

File: Week 6/Pset6/sentimental-readability/readability.py
import sys
from cs50 import get_string
import re


def get_letter_count(text):
    # Counts number of letters in text, using built in method .isalpha()
    letter_count = 0
    for i in text:
        if i.isalpha():
            letter_count += 1
    return letter_count


def get_word_count(text):
    # Count words in text. Spaces and end of text will count as a delimiter for words
    word_count = 0
    for i, letter in enumerate(text):
        if letter.isspace() or (len(text) != 0 and i == len(text) - 1):
            word_count += 1
    return word_count


def get_sentence_count(text):
    # Count sentences in text. [!.?] are used as sentences delimiters
    sentence_count = 0
    for i, letter in enumerate(text):
        if letter == "!" or letter == "." or letter == "?":
            sentence_count += 1
    return sentence_count


def calculate_score(avg_letters, avg_sentences):
    # Calculates text score based on Coleman-Liau
    return round(0.0588 * avg_letters - 0.296 * avg_sentences - 15.8)


def get_grade(score):
    # Evaluates Coleman-Liau score to grade
    if (score > 16):
        return "Grade 16+"
    elif (score >= 1 and score <= 16):
        return "Grade {}".format(score)
    else:
        return "Before Grade 1"


def get_hundred_word_avg(count, nr_of_words):
    return count / nr_of_words * 100.0


def get_grade_for_text(text):
    # Find counts required for the score calculation
    letter_count = get_letter_count(text)
    word_count = get_word_count(text)
    sentence_count = get_sentence_count(text)
    #  Calculate 100 word average counts
    letter_hundred_avg = get_hundred_word_avg(letter_count, word_count)
    sentences_hundred_avg = get_hundred_word_avg(sentence_count, word_count)
    # Calculate score based on Coleman-Liau
    score = calculate_score(letter_hundred_avg, sentences_hundred_avg)
    return get_grade(score)


def main():
    input_text = get_string("Text: ")
    print(get_grade_for_text(input_text))


if __name__ == "__main__":
    main()
    sys.exit(0)
